src/librustdoc/doctest.rs

```diff
@@ -244,16 +250,87 @@ pub(crate) fn run(dcx: DiagCtxtHandle<'_>, options: RustdocOptions) -> Result<()
+    opts: GlobalTestOptions,
+    rustdoc_options: &Arc<RustdocOptions>,
+    unused_extern_reports: &Arc<Mutex<Vec<UnusedExterns>>>,
+    mut standalone_tests: Vec<test::TestDescAndFn>,
+    mergeable_tests: FxHashMap<Edition, Vec<(DocTestBuilder, ScrapedDoctest)>>,
 ) {
+    let mut test_args = Vec::with_capacity(rustdoc_options.test_args.len() + 1);
     test_args.insert(0, "rustdoctest".to_string());
-    if nocapture {
+    test_args.extend_from_slice(&rustdoc_options.test_args);
+    if rustdoc_options.nocapture {
         test_args.push("--nocapture".to_string());
     }
-    tests.sort_by(|a, b| a.desc.name.as_slice().cmp(&b.desc.name.as_slice()));
-    test::test_main(&test_args, tests, None);
+
+    let mut nb_errors = 0;
+    let mut ran_edition_tests = 0;
+    let target_str = rustdoc_options.target.to_string();
+
+    for (edition, mut doctests) in mergeable_tests {
+        if doctests.is_empty() {
+            continue;
+        }
+        // If there is only one mergeable doctest, the cost to run it would be higher than just
+        // running it alonside standalone doctests.
+        if doctests.len() > 1 {
+            doctests.sort_by(|(_, a), (_, b)| a.name.cmp(&b.name));
+
+            let mut tests_runner = runner::DocTestRunner::new();
+
+            let rustdoc_test_options = IndividualTestOptions::new(
+                &rustdoc_options,
+                &Some(format!("merged_doctest_{edition}")),
+                PathBuf::from(format!("doctest_{edition}.rs")),
+            );
+
+            for (doctest, scraped_test) in &doctests {
+                tests_runner.add_test(doctest, scraped_test, &target_str);
+            }
+            if let Ok(success) = tests_runner.run_merged_tests(
+                rustdoc_test_options,
+                edition,
+                &opts,
+                &test_args,
+                rustdoc_options,
+            ) {
+                ran_edition_tests += 1;
+                if !success {
+                    nb_errors += 1;
+                }
+                continue;
+            }
+            // We failed to compile all compatible tests as one so we push them into the
+            // `standalone_tests` doctests.
+            debug!("Failed to compile compatible doctests for edition {} all at once", edition);
+        }
+        for (doctest, scraped_test) in doctests {
+            doctest.generate_unique_doctest(
+                &scraped_test.text,
+                scraped_test.langstr.test_harness,
+                &opts,
+                Some(&opts.crate_name),
+            );
+            standalone_tests.push(generate_test_desc_and_fn(
+                doctest,
+                scraped_test,
+                opts.clone(),
+                Arc::clone(&rustdoc_options),
+                unused_extern_reports.clone(),
+            ));
+        }
+    }
+
+    // We need to call `test_main` even if there is no doctest to run to get the output
+    // `running 0 tests...`.
```

Comment:
I thought it was important to make an output because in this case, the doctests will be run in the same process, which can alter the results.

Reply:
I'm just saying it's confusing to output "running 0 tests..." what does that mean exactly? It means all tests were merged? In that case, we should say something like that instead.

Reply:
Ah no, it doesn't display that if there are 0 doctests. It displays in case there were no merged doctests and no "standalone" doctests to keep the current behaviour.